networking_ccloud/ml2/agent/common/switch.py

```diff
@@ -129,10 +182,29 @@ def persist_config(self):
     def _persist_config(self):
         raise NotImplementedError
 
-    @run_in_executor('write')
     def run_full_sync(self, context):
-        """Schedule a full sync for this switch. Config will be fetched when the sync starts"""
-        self._run_full_sync(context)
+        """Schedule a full sync for this switch or re-use an already scheduled one and return its future.
+
+        Config will be fetched when the sync starts
+        """
+
+        @run_in_executor('write')
+        def _in_thread_full_sync(self, context):
+            # once the full sync starts, we remove access to the future,
+            # because new state coming in might not be seen by a running
+            # full-sync anymore and thus be missed
+            with self._full_sync_future_lock:
+                self._full_sync_future = None
+
+            return self._run_full_sync(context)
```

Comment:
With this we reset the full sync future before actually running the sync, so we miss out on catching everyone that joins the party™ while the full sync runs. Wouldn't it make more sense to reset the full sync future once we have the result (or an exception) of `_run_full_sync()`, but before returning (i.e. concluding the future)? Could this be a case for `try`/`finally`?

Reply:
Ah, I just now have "been read" the commit message. We want to make sure we don't loose any updates and that could happen, because updates _after_ fetching the config still need to be applied. Totally makes sense!